### PR TITLE
[Plugin]: MSW GraphQL mocks

### PR DESCRIPTION
**Pull request by smhmd**
Fixes: #6924

## Description

This PR adds a plugin for generating [MSW](https://mswjs.io/) GraphQL mock handlers.
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Given this input in the first code block, we'll generate the output in the second codeblock:

Operation:
```graphql
query PullRequest ($id: Int!) {
  pr(id: $id) {
    approved
  }
}
```

The variables and selection in `req.variabels` and `ctx.data` are taken from the operations to generate useful JSDoc comments:

```ts
import { graphql, ResponseResolver, GraphQLRequest, GraphQLContext } from 'msw'

/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockPullRequestQuery((req, res, ctx) => {
 *   const { id } = req.variables;
 *   return res(
 *     ctx.data({ pr })
 *   )
 * })
 */
export const mockPullRequestQuery = (resolver: ResponseResolver<GraphQLRequest<PullRequestQueryVariables>, GraphQLContext<PullRequestQuery>, any>) =>
  graphql.query<PullRequestQuery, PullRequestQueryVariables>(
    'PullRequest',
    resolver
  )
```

Usage:
```ts
import { mockPullRequestQuery } from '../api'

const worker = setupWorker(
  mockPullRequestQuery((req, res, ctx) => res(ctx.data({ pr: { approved: false } }))),
)

worker.start()
```

Only mutations and queries are handled. They generate functions that are named: `mock<OperationName><OperationType><EndpointName>` (`mockUserQuery` or `mockAdminMutationApi2`, for instance).

You can also add a `link: { name: string, endpoint: string }` option to generate the mock handlers using a [link](https://mswjs.io/docs/api/graphql/link).

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Tests

- [x] A test to see if the plugin generates the handler correctly.
- [x] A test to see if the plugin handles a "link" configuration correctly.

Run the tests by running `yarn test -t 'msw'`.

**Test Environment**:
- OS: Linux
- `@graphql-codegen/typescript-msw`: 0.0.0
- NodeJS: v16.11.0

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/plugins/typescript/msw/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# @graphql-codegen/typescript-msw
+
+## 1.0.0
+
+### Major Changes
+
+- First release of `typescript-msw` plugin.
+  Generates MSW mock handlers based on GraphQL operations
```
Comment (n1ru4l) [inline packages/plugins/typescript/msw/CHANGELOG.md:8]: We are using changesets for versioning. Please remove the changelog and instead create a changeset as described in https://github.com/dotansimha/graphql-code-generator/pull/6945#issuecomment-955581309

File: packages/plugins/typescript/msw/tests/msw.spec.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    expect(result.content).toContain(
+      `export const updateUserMutationApiHandler = (resolver: ResponseResolver<GraphQLRequest<UpdateUserMutationVariables>, GraphQLContext<UpdateUserMutation>, any>) =>
+  api.mutation<UpdateUserMutation, UpdateUserMutationVariables>(
+    'UpdateUser',
+    resolver
+  )`
+    );
+  });
+});
```
Comment (n1ru4l) [inline packages/plugins/typescript/msw/tests/msw.spec.ts:41]: Can you please extend this test suite to cover some more cases? Two test cases seem few for a new plugin. Also there is a `it.only` on one of the tests

**Comment by n1ru4l** [comment]
Thank you for starting this new plugin! Could you also add documentation with usage examples, so people can understand how and why they should adopt this plugin?

I also added a few comments

**Comment by dotansimha** [comment]
@smhmd ^ ping :)

File: packages/plugins/typescript/msw/tests/msw.spec.ts (hunk 2 above)
Comment (smhmd) [reply packages/plugins/typescript/msw/tests/msw.spec.ts]: The `it.only` was removed 👍🏼 . Perhaps, adding [`no-only-tests` ESLint rule](https://www.npmjs.com/package/eslint-plugin-no-only-tests) would help.

I added a few assertions and one test for the generated JSDoc. Let me know if you have something else to test in mind. The plugin is very simple.

**Comment by smhmd** [comment]
@n1ru4l @dotansimha Documentation was added and the plugin was altered a little bit. Please, recheck the PR description.
Changes are:
- Generated functions have different name structure. (We follow MSW documentation examples, now.)
- Added JSDocs to generated functions.

File: packages/plugins/typescript/msw/tests/msw.spec.ts (hunk 2 above)
Comment (n1ru4l) [reply packages/plugins/typescript/msw/tests/msw.spec.ts]: @smhmd Feel free to send a PR that sets this up 💯

**Review by n1ru4l — APPROVED** [review]
@smhmd Thank you :) Feel free to follow up with a PR for https://www.npmjs.com/package/eslint-plugin-no-only-tests

**Pull request merged**